### DISCUSSION TITLE
Custom Directive on INPUT_FIELD_DEFINITION

### DISCUSSION THREAD

**Discussion by maxixcom** [Netflix/dgs-framework#594] [Q&A]
Hello everyone! 

The idea is having custom directive to filter input object fields if user haven't enough permissions to change.

Here is the schema:

```graphql
directive @hasPermission(permission: String!) on INPUT_FIELD_DEFINITION

type Mutation {
    """
    Update user
    """
    userUpdate(input: UserUpdateInput!) : UserUpdatePayload
}

"""
Use this object to pass data to update user
"""
input UserUpdateInput {
    id: ID!
    firstName: String
    lastName: String
    roles: [UserRole!] @hasPermission(permission: "admin:write")
    status: UserStatus
}

"""
Updated user result
"""
type UserUpdatePayload {
    id: ID!
    record: User
}

"""
Represents User object
"""
type User {
    id: ID!
    email: String
}
```

I tried this approach https://github.com/Netflix/dgs-framework/discussions/69#discussioncomment-386297
and added child datafetcher for `UserUpdateInput` input object:

```java
    @DgsData(parentType = "Mutation", field = "userUpdate")
    fun userUpdate(
        @Valid @InputArgument("input") input: UserUpdateInput,
        dfe: DgsDataFetchingEnvironment
    ): CompletableFuture<UserUpdatePayload> = scope.future {
        dfe.getDgsContext().checkAuthority(UserPermission.ManagerWrite.permission)

        userService.userUpdate(input)
    }

    @DgsData(parentType = "UserUpdateInput", field = "roles")
    fun userUpdateInputRoles(
        dfe: DataFetchingEnvironment
    ): List<String>? {
        val hasPermission = dfe.fieldDefinition.getDirective("hasPermission")
        val permission = hasPermission.getArgument("permission").value

        return null
    }
```

But `userUpdateInputRoles` child datafetcher have never been called. Is it a bug or I did something wrong?

Could anyone help me please with that? 
Thanks

**Comment by manasiy** [comment]
@maxixcom 
I am looking at implementing something similar.
How did you fix it ? Do you have a working example for this use case ?
Thank you

**Reply by srinivasankavitha** [reply]
The above example is incorrect since the data fetcher `userUpdateInputRoles` will not be called on input fields. You need to extract the input fields from `dfe` in the parent data fetcher  `userUpdate` and apply any logic for your directive there.
